Source/WebCore/svg/SVGLengthValue.cpp

```diff
@@ -309,39 +285,143 @@ float SVGLengthValue::value(const SVGLengthContext& context) const
     return result.releaseReturnValue();
 }
 
+float SVGLengthValue::valueAsPercentage() const
+{
+    return WTF::switchOn(m_value,
+        [](const CSS::Number<>& number) -> float {
+            if (auto raw = number.raw())
+                return raw->value;
+
+            return 0.0f;
+        },
+        [](const CSS::LengthPercentage<>& length) -> float {
+            if (auto raw = length.raw()) {
+                if (raw->unit == CSS::LengthPercentageUnit::Percentage)
+                    return raw->value / 100.0f;
+
+                return raw->value;
+            }
+
+            return 0.0f;
+        }
+    );
+}
+
+float SVGLengthValue::valueInSpecifiedUnits() const
+{
+    // Per SVG spec: return 0 for non-scalar values like calc()
+    // https://svgwg.org/svg2-draft/types.html#__svg__SVGLength__valueInSpecifiedUnits
+
+    return WTF::switchOn(m_value,
+        [](const CSS::Number<>& number) {
+            if (auto raw = number.raw())
+                return clampTo<float>(raw->value);
+
+            return 0.f;
+        },
+        [](const CSS::LengthPercentage<>& length) {
+            if (length.isCalc())
+                return 0.f;
+
+            if (auto raw = length.raw())
+                return clampTo<float>(raw->value);
+
+            return 0.f;
+        }
+    );
+}
+
 String SVGLengthValue::valueAsString() const
 {
-    return makeString(m_valueInSpecifiedUnits, lengthTypeToString(m_lengthType));
+    return WTF::switchOn(m_value,
+        [](const CSS::Number<>& number) {
```

Comment:
Seems like we should use `const auto&` here to avoid the copied and pasted code.